### ISSUE TITLE
Question: use of setdiff in unique_names

### ISSUE THREAD

**Issue by imanuelcostigan** [hadley/dplyr#709]
Is there a reason why [this](https://github.com/hadley/dplyr/blame/6a1e90830d3c659874cb6d3fd7f96e722447882e/R/join.r#L113) is wrapped with a `setdiff` in `unique_names()`:

```R
common <- setdiff(intersect(x_names, y_names), by)
```

This causes issues for SQL Server as it doesn't have natural joins (which only returns one copy of attributes that are part of the join predicate). In particular, if the join predicate is tbl1.id = tbl2.id, then SQL Server will not attempt to drop one tables' id attribute...it will return both and the resulting data set will have two columns with same name. 

This has consequences because `sql_join()` returns a query of the form select * from (<Actual join SQL>) which will fail because the `*` will try to select two fields with the same name (in this case, `id`).

**Comment by piccolbo** [comment]
+1 for hiveQL

**Comment by hadley** [comment]
Could you please provide a small reprex that I can use to test I don't break existing postgres/SQLite behaviour?

**Comment by imanuelcostigan** [comment]
Taking a look at the dev version of dplyr, this is not only going to persist, but will require a huge amount of work (and dplyr code duplication) on my backend (the whole ops/sql_build/sql_render stack AFAICT). I am conscious that I owe you a reprex, but not quite sure how I can give you more than what I have pasted above from SQL Server side - and this issue doesn't affect the backends you mention. Are you asking for an example of how generated code changes for SQLite / Postgres were I to change `unique_names()` per my comment above?

**Comment by hadley** [comment]
It looks like this change will affect the column names generated from any query with duplicate column names: `"fieldA", "fieldB", "commonField", "fieldC"` -> `"fieldA", "fieldB", "commonField.x", "fieldC"`, and given that except when I'm deeply immersed in dplyr I don't understand the ramification of this sort of change, you need to help me understand what's going on. Helping me understand includes making the simplest possible reprex that I can run and think about.

**Comment by imanuelcostigan** [comment]
Reprex as requested (I hope): http://rpubs.com/imanuel/191061

Uses: imanuelcostigan/dplyr@d9314ec

**Comment by imanuelcostigan** [comment]
### Before change

```R
library(dplyr) # CRAN v0.4.3
src <- src_sqlite(tempfile(), create = TRUE)
data_frame(a = 1:3, b = letters[1:3]) %>%
  copy_to(src, ., "df1")
data_frame(b = letters[1:3], c = letters[24:26]) %>%
  copy_to(src, ., "df2")
left_join(tbl(src, "df1"), tbl(src, "df2"))
```

Result: 
```
### df1 
Source:   query [?? x 2]
Database: sqlite 3.8.6 [/var/folders/9f/kz67r0rn3qb9tx_rmgbps94r0000gq/T//RtmpNgdXFg/file1266f44831969]

       a     b
   <int> <chr>
1      1     a
2      2     b
3      3     c
..   ...   ...

### df2

Source:   query [?? x 2]
Database: sqlite 3.8.6 [/var/folders/9f/kz67r0rn3qb9tx_rmgbps94r0000gq/T//RtmpNgdXFg/file1266f44831969]

       b     c
   <chr> <chr>
1      a     x
2      b     y
3      c     z
..   ...   ...


### left_join(df1, df2)
Source: sqlite 3.8.6 [/var/folders/9f/kz67r0rn3qb9tx_rmgbps94r0000gq/T//Rtmpw2oVTr/file1263c7a25c52b]
From: <derived table> [?? x 3]

       a     b     c
   (int) (chr) (chr)
1      1     a     x
2      2     b     y
3      3     c     z
..   ...   ...   ...
```

### After change

See [diff](https://github.com/imanuelcostigan/dplyr/commit/d9314eca57813aa4de933be269aa00bde9b2a549).  Restart your R session.

```R
dev_mode()
# install_github("imanuelcostigan/dplyr@d9314ec")
library(dplyr)
src <- src_sqlite(tempfile(), create = TRUE)
data_frame(a = 1:3, b = letters[1:3]) %>%
  copy_to(src, ., "df1")
data_frame(b = letters[1:3], c = letters[24:26]) %>%
  copy_to(src, ., "df2")
left_join(tbl(src, "df1"), tbl(src, "df2"))
```

Result

```
### left_join(df1, df2)

Source:   query [?? x 4]
Database: sqlite 3.8.6 [/var/folders/9f/kz67r0rn3qb9tx_rmgbps94r0000gq/T//RtmpNgdXFg/file1266f44831969]

       a   b.x   b.y     c
   <int> <chr> <chr> <chr>
1      1     a     a     x
2      2     b     b     y
3      3     c     c     z
..   ...   ...   ...   ...
```

**Comment by imanuelcostigan** [comment]
This issue would probably be resolved by following [comment based TODO](https://github.com/hadley/dplyr/blob/563b5629d1680d6065c84f85501aed609838b881/R/sql-build.R#L180) being closed:

>  # TODO: it would be better to construct an explicit FROM statement
>  # that used the table names to disambiguate the fields names: this
>  # would remove a layer of subqueries and would make sql_join more
>  # flexible.

**Comment by hadley** [comment]
@javierluraschi is going to look into this as we need it for spark SQL

**Comment by imanuelcostigan** [comment]
What was be neat is if in the left / right join process only the .x / .y version of field is returned without the .x / .y suffix.

**Comment by javierluraschi** [comment]
@imanuelcostigan Could you take a look at this PR?

From your natural join example, this PR will this result:

```
<SQL> SELECT * FROM 'df1'

LEFT JOIN

'df2'

USING ('b')
```

into the following SQL statement:

```
SELECT 'a', 'c', 'b.x' AS 'b'
FROM (
   SELECT * FROM 
      (SELECT 'a' AS 'a', 'b' AS 'b.x' FROM 'df1')
   LEFT JOIN
      (SELECT 'b' AS 'b.y', 'c' AS 'c' FROM 'df2')
   ON ('b.x' = 'b.y')
)
```

would be great if you could sync to this change and validate this is fixed as you would expect.

**Comment by imanuelcostigan** [comment]
Closed up #1997
